README.md
# PWDS #
- - - -
## What is PWDS ##
* PWDS is a local, command line encrypted password storage system with different actions and prefernces avaliable 
* The entire program started with a little summer project and I never thought it would develop into a large-scale project :smiling_imp:
* The whole program is built on python3
- - - -
## How does it work ##
* PWDS creates a database and allows you to change different attributes of it.
* Regular backup will be done.
* Cleverly uses clear screen function so that your passwords will not be kept on the terminal screen
* Even if you mess up, there will be a backup to save your day! Import backups are now avaliable!  
- - - -  
## Possible problems you may encounter ##  
### WrongPasswordError ###  
* As its name suggests, you probably entered the wrong password. 
* As for extra security, we do not allow 'try agains' ~~we are just being mean~~
* The occurance of wrong password entering will not affect your database! However, we will record every single attempt to login if 'Log login' preference is set to true

### EncryptionError ###
* As of today (12/14/2019), we only have 128 characters avaliable for password.
* Any character used outside of these 128 characters will not be encrypted and will raise encryption error

- - - -
## Exports ##
* Sounds strange but we do let you export your saved password and logs in both encrypted and non-encrypted form.  
Avaliable output format includes:  

.txt|.csv|.json| .db  
---- | ----| ---- | ----
Plain text file | Comma Seperated Values file | JavaScript Object Notation | SQL Database FIle   


changelog.md
# Change Logs #
- - - -
### 2019/12/15 04:25 ###
#### Changed License.md to my actual name ####

#### Added [wait for input] into funcs.py  ####
now it imports a class and prints the color imported  

#### Added new file userClass.py ####
pwd.py imports a class from userClass.py
neater codes  

#### Added pwd.py previous class structure into userClass.py ####
pwd.py now only includes prompt and calling class
neater codes, easier references  
- - - -
### 2019/12/16 01:48 ###
#### New procedure in login function ####
instantQuit skips userInterface.checkbackup() function and enhances performance  
#### Altered pwd.py login process ####
if newuser: {initialise...login}else: {login}-->if newuser: {initialise} {login}  
#### Added error handling 'instantQuit' in pwd.py  ####
New error handling prevents possible flaw while importing backups.  
- - - -
### 2019/12/16 08:49 ###
#### introduce delete function ####
Deletes item from database `still in development`
- - - -  
### 2019/12/16 11:50 ###
#### Changed encryption method #### 
Password is now hashed and hexified before encrypting to prevent injection attacks  
improved 'Delete' function `still in development`
- - - -
### 2019/12/16 15:24 ###
#### Finished function delete ####
Finally. Introducing function 'delete' from database
slight alteration in var types
- - - -
### 2019/12/16 16:33 ###
#### Edited encryption ####
Changed password hashing from 'sha256' to 'sha512'
- - - -
### 2019/12/16 22:11 ###
#### A reform in database storage ####
All 'key' values inside the database is now slightly encrypted //symmetrical  
logs are includeed  
#### Encryption of keys
password hashing now uses `sha512(password) + sha512(''.join(sorted(password, reversed=True)))`  
longer password allows higher security  
- - - -
### 2019/12/17 16:30 ###
#### New encryption method ####
Today was a busy day (even though I have my finals right now)  
Found a major flaw in 'enc.py' causing encryption problems, completely redesigned the encrypting alogrithm
_Now supports any character!!_  
#### More hashing ####
Anything that includes the process of using your password now uses hashed password (of course)  
No functionality addition  
### 2019/12/20 12:07 ###
It's been a few days since I last commited cus I have my finals  
just finished my finals so the update should be faster and its holiday I mean...  
#### Modified user preference function ####
Still working on it. String formatting is kind of killing me but I should be able to figure it out soon
Output looks fine
The past few days I changed the 'colors' returning method, making it a bit faster I believe  
#### Fixed encryption method! ####
I also fixed the 'last character problem' in enc.py, eliminating possible errors.
### 2019/12/24 00:10 ###
Whoa it's been a few days since my last push.  
Spent my weekend going around killing time...did't work a whole lot on it lol  
Added many many more functionality to option `verbose` 
New file alterPrefs. Used to change preferences `still in development`  

exec/enc.py
# Matthew Lam
# module designed to import to [[encrypt]] and [[decrypt]]
import random, hashlib

newd = {
	' ' : 0,
	'a' : 1,
	'b' : 2,
	'c' : 3,
	'd' : 4,
	'e' : 5,
	'f' : 6,
	'g' : 7,
	'h' : 8,
	'i' : 9,
	'j' : 10,
	'k' : 11,
	'l' : 12,
	'm' : 13,
	'n' : 14,
	'o' : 15,
	'p' : 16,
	'q' : 17,
	'r' : 18,
	's' : 19,
	't' : 20,
	'u' : 21,
	'v' : 22,
	'w' : 23,
	'x' : 24,
	'y' : 25,
	'z' : 26,
	'A' : 27,
	'B' : 28,
	'C' : 29,
	'D' : 30,
	'E' : 31,
	'F' : 32,
	'G' : 33,
	'H' : 34,
	'I' : 35,
	'J' : 36,
	'K' : 37,
	'L' : 38,
	'M' : 39,
	'N' : 40,
	'O' : 41,
	'P' : 42,
	'Q' : 43,
	'R' : 44,
	'S' : 45,
	'T' : 46,
	'U' : 47,
	'V' : 48,
	'W' : 49,
	'X' : 50,
	'Y' : 51,
	'Z' : 52,
	',' : 53,
	'<' : 54,
	'.' : 55,
	'>' : 56,
	'/' : 57,
	'?' : 58,
	';' : 59,
	':' : 60,
	'[' : 61,
	'{' : 62,
	']' : 63,
	'}' : 64,
	'|' : 65,
	'`' : 66,
	'~' : 67,
	'1' : 68,
	'!' : 69,
	'2' : 70,
	'@' : 71,
	'3' : 72,
	'#' : 73,
	'4' : 74,
	'$' : 75,
	'5' : 76,
	'%' : 77,
	'6' : 78,
	'^' : 79,
	'7' : 80,
	'&' : 81,
	'8' : 82,
	'*' : 83,
	'9' : 84,
	'(' : 85,
	'0' : 86,
	'-' : 87,
	'_' : 88,
	'=' : 89,
	'+' : 90,
	'å' : 91,
	'∫' : 92,
	'∂' : 93,
	'©' : 94,
	'†' : 95,
	'π' : 96,
	'"' : 97,
	'\'' : 98,
	')' : 99,}

newdec ={v: k for k, v in newd.items()}

encd = {
	' ' : 0,
	'a' : 1,
	'b' : 2,
	'c' : 3,
	'd' : 4,
	'e' : 5,
	'f' : 6,
	'g' : 7,
	'h' : 8,
	'i' : 9,
	'j' : 10,
	'k' : 11,
	'l' : 12,
	'm' : 13,
	'n' : 14,
	'o' : 15,
	'p' : 16,
	'q' : 17,
	'r' : 18,
	's' : 19,
	't' : 20,
	'u' : 21,
	'v' : 22,
	'w' : 23,
	'x' : 24,
	'y' : 25,
	'z' : 26,
	'A' : 27,
	'B' : 28,
	'C' : 29,
	'D' : 30,
	'E' : 31,
	'F' : 32,
	'G' : 33,
	'H' : 34,
	'I' : 35,
	'J' : 36,
	'K' : 37,
	'L' : 38,
	'M' : 39,
	'N' : 40,
	'O' : 41,
	'P' : 42,
	'Q' : 43,
	'R' : 44,
	'S' : 45,
	'T' : 46,
	'U' : 47,
	'V' : 48,
	'W' : 49,
	'X' : 50,
	'Y' : 51,
	'Z' : 52,
	',' : 53,
	'<' : 54,
	'.' : 55,
	'>' : 56,
	'/' : 57,
	'?' : 58,
	';' : 59,
	':' : 60,
	'[' : 61,
	'{' : 62,
	']' : 63,
	'}' : 64,
	'|' : 65,
	'`' : 66,
	'~' : 67,
	'1' : 68,
	'!' : 69,
	'2' : 70,
	'@' : 71,
	'3' : 72,
	'#' : 73,
	'4' : 74,
	'$' : 75,
	'5' : 76,
	'%' : 77,
	'6' : 78,
	'^' : 79,
	'7' : 80,
	'&' : 81,
	'8' : 82,
	'*' : 83,
	'9' : 84,
	'(' : 85,
	'0' : 86,
	'-' : 87,
	'_' : 88,
	'=' : 89,
	'+' : 90,
	'å' : 91,
	'∫' : 92,
	'∂' : 93,
	'©' : 94,
	'†' : 95,
	'π' : 96,
	'"' : 97,
	'…' : 98,
	')' : 99,
	'∞' : 100,
	'∏' : 101,
	'Í' : 102,
	'◊' : 103,
	'¬' : 104,
	'€' : 105,
	'¿' : 106,
	'Â' : 107,
	'º' : 108,
	'÷' : 109,
	'«' : 110,
	'Æ' : 111,
	'°' : 112,
	'‹' : 113,
	'ø' : 114,
	'ˇ' : 115,
	'‡' : 116,
	'·' : 117,
	'»' : 118,
	'Ú' : 119,
	'√' : 120,
	'∆' : 121,
	'\'' : 122,
	'Ç' : 123,
	'Î' : 124,
	'Ô' : 125,
	'Œ' : 126,
	'ﬂ' : 127,
	'§' : 128}

decd = {v: k for k, v in encd.items()}

def encsp(string, password):
	inv = {'0': '1', '1': '0'}
	keyd = {
	0: 0,
	1:(0,1),
	2:(0,2),
	3:(0,3),
	4:(0,4),
	5:(0,5),
	6:(0,6),
	7:(1,2),
	8:(1,3),
	9:(1,4),
	10:(1,5),
	11:(1,6),
	12:(2,3),
	13:(2,4),
	14:(2,5),
	15:(2,6),
	16:(3,4),
	17:(3,5),
	18:(3,6),
	19:(4,5),
	20:(4,6),
	21:(5,6),
	22:1,
	23:2,
	24:3,
	25:4,
	26:5,
	27:6}
	key = ((len(str(password))+len(str(string))) % 28)
	spbin = ["{0:b}".format(int(encd.get(chars))).zfill(7) for chars in string]
	ilist = []
	for items in spbin:
		spstr = ''
		for count, bins in enumerate(items):
			spstr = spstr + (bins if str(count) in str(keyd.get(key)) else inv.get(str(bins)))
		ilist.append(spstr)
	return ''.join([decd.get(int(str(i), 2)) for i in ilist])

def encrypt(encs, pwd):
	encs = ''.join(str(ord(chars)).zfill(7) for chars in str(encs))
	front0s = [encs.index(i) for i in encs if i != '0'][0]
	ending0s = len(encs) - len(encs.rstrip('0'))
	npwd = int(hashlib.pbkdf2_hmac('sha512', str(pwd).encode('utf-32'), ''.join(sorted(pwd, reverse=True)).encode('utf-32'), 300000).hex(), 16)
	intencd = str(int(encs) * int(npwd))
	cy = ''.join(random.choices([m for m in newdec.values()], k=4)) + str(front0s).zfill(2) + str(ending0s).zfill(2)
	for p in [int(intencd[i:i+2]) for i in range(0, len(intencd), 2)]:
		cy += newdec.get(p)
	return encsp(cy, pwd)

def decrypt(decs, pwd):
	decs = encsp(decs, pwd)
	zeros = decs[4:8]
	product = ''.join([str(newd.get(i)).zfill(2) for i in decs[8:]])
	npwd = int(hashlib.pbkdf2_hmac('sha512', str(pwd).encode('utf-32'), ''.join(sorted(pwd, reverse=True)).encode('utf-32'), 300000).hex(), 16)
	divided = str(int(product) // int(npwd))
	subzeroed = ('0' * int(zeros[:2]) if int(zeros[:2]) > 0 else '') + divided + ('0' * int(zeros[2:]) if int(zeros[2:])> 0 else '')
	subzlist = [int(subzeroed[i:i+7]) for i in range(0,len(subzeroed), 7)]
	if subzlist[-1] <10 and subzlist[-1]>0:
		subzlist[-2] += 1
		subzlist.pop(-1)
	elif subzlist[-1] == 0:
		subzlist.pop(-1)
	return ''.join([chr(k) for k in subzlist])


















